sharktank/sharktank/layers/ffn_moe_block.py

```diff
@@ -74,7 +77,105 @@ def forward(
         return torch.sum(ffn_down, dim=1)
 
 
-class FFNMOE(ThetaLayer):
+class DenseFFNMOE(ThetaLayer):
+    """Do not route the tokens to only the top selected experts, but route to all
+    experts.
+
+    This would result in doing useless computation for some token expert pairs where
+    the routing score is 0, but we would perform fewer number matrix multiplication.
+    What approach is better would depend on the number experts and the number of top
+    experts.
+    """
+
+    def __init__(
+        self,
+        theta: Theta,
+        rms_epsilon: float | None = None,
+        is_gated: bool = True,
+        activation_fn: Callable[[torch.Tensor], torch.Tensor] = F.silu,
+        activation_dtype: Optional[torch.dtype] = None,
+        fake_quant: bool = False,
+    ):
+        super().__init__(theta)
+        self.num_experts = theta("ffn_gate_exps", "weight").shape[0]
+        ffn_theta = Theta(
+            {
+                "ffn_gate": theta("ffn_gate_exps").tree,
+                "ffn_up": theta("ffn_up_exps").tree,
+                "ffn_down": theta("ffn_down_exps").tree,
+            }
+        )
+        self.ffn = FFN(
+            ffn_theta,
+            rms_epsilon=rms_epsilon,
+            is_gated=is_gated,
+            activation_fn=activation_fn,
+            activation_dtype=activation_dtype,
+            fake_quant=fake_quant,
+            add_residual=False,
+        )
+
+    def forward(
+        self,
+        h: torch.Tensor,
+        top_experts_index: torch.Tensor,
+        expert_gate: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        h:
+            Input tokens.
+            Shape of (batch_size * sequence_length, input_feature_dim).
+        top_experts_index:
+            indexes of selected top experts for each token.
+            Shape of (batch_size * sequence_length, num_top_experts).
+            Value range is [0, num_of_experts)
+        expert_gate:
+            router weights of selected top experts for each token.
+            Shape of (batch_size * sequence_length, num_top_experts).
+
+        Returns:
+            Tensor of shape (batch_size * sequence_length, expert_feature_dim)
+        """
+        num_tokens, input_feature_dim = h.shape
+
+        # (self.num_experts, num_tokens)
+        router_scores = (
+            torch.full([num_tokens, self.num_experts], 0, dtype=h.dtype)
+            .scatter_(1, top_experts_index, expert_gate)
+            .transpose(0, 1)
+        )
+
+        # (self.num_experts, num_tokens)
+        router_indices = (
+            torch.arange(num_tokens, device=h.device)
```

Comment:
We should be careful about using `torch` operators explicitly. It causes problems when we try to do sharding work. Long term we should have `range` as an operation in sharktank's ops.